File: perf/Perf.cs
using GosharpTemplate;
using System.Diagnostics;

public class Perf
{
    const bool RUN_GO = true; 
    const int RUNS = 1000;

    public static void Main(params string[] args)
    {
        Console.WriteLine("Run Csharp:");
        Table1000Items();

#if true
        Console.WriteLine();
        Console.WriteLine();
        Console.WriteLine("Run Go:");
        Process cmd = new Process();
        cmd.StartInfo.FileName = "cmd.exe";
        cmd.StartInfo.RedirectStandardInput = true;
        cmd.StartInfo.RedirectStandardOutput = true;
        cmd.StartInfo.CreateNoWindow = true;
        cmd.StartInfo.UseShellExecute = false;
        cmd.Start();

        cmd.StandardInput.WriteLine("go run ../../../perf.go");
        cmd.StandardInput.Flush();
        cmd.StandardInput.Close();
        cmd.WaitForExit();
        Console.WriteLine(cmd.StandardOutput.ReadToEnd());
#endif
    }

    static void Table1000Items()
    {
        //Setup
        var template = new Template();
        //template.ParseFiles("List.html");
        template.ParseFiles("../../../List.html");
        var items = GenerateRange(1000);
        var sw = new Stopwatch();

        var times = new List<double>(RUNS);
        for (int i = 0; i < RUNS; i++)
        {
            sw.Start();
            template.ExecuteTemplate("table", items);
            sw.Stop();
            times.Add(sw.Elapsed.TotalMicroseconds);
            sw.Reset();
        }
        Console.WriteLine($"Table with 1000 items, {RUNS} runs:");
        Console.WriteLine($"Min: {times.Skip(1).Min()} (micros)");
        Console.WriteLine($"Max: {times.Skip(1).Max()} (micros)");
        Console.WriteLine($"Average: {times.Skip(1).Average()} (micros)");
    }

    record Customer(
        string Name,
        string Address,
        string City,
        string Zip,
        string EMail,
        string Phone
    );

    record TableItem(
        int Id,
        string Name,
        bool Valid,
        string Description,
        int Count,
        double Price,
        Customer Customer
    );

    static List<TableItem> GenerateRange(int n)
    {
        var items = new List<TableItem>(n);
        for (int i = 1; i <= n; i++)
        {
            items.Add(
                new TableItem(
                    i,
                    $"Item{i}",
                    true,
                    "Description {i}",
                    i % 100,
                    Random.Shared.NextDouble() * 1000.0,
                    new Customer(
                        $"Customer {i}",
                        $"Customer Street {i}",
                        $"City{i}",
                        $"{i}",
                        $"Cust.omer{i}@email.com",
                        $"{i}"
                    )
                )
            );
        }
        return items;
    }
}

File: perf/perf.go
package main

import (
	"bytes"
	"fmt"
	"text/template"
	"time"
)

const RUNS = 1000

type TableItem struct {
	Id          int
	Name        string
	Valid       bool
	Description string
	Count       int
	Price       float64
	Customer	Customer
}

type Customer struct {
	Name	string
	Address string
	City	string
	Zip		string
	EMail	string
	Phone	string
}	


func GenerateRange(n int) []TableItem {
	var items = make([]TableItem, n)
	for i := 0; i < n; i++ {
		items[i] = TableItem{
			i,
			fmt.Sprintf("item %d", i),
			true,
			fmt.Sprintf("Description %d", i),
			i % 100,
			100.01,
			Customer {
				fmt.Sprintf("Customer %d", i),
				fmt.Sprintf("Customer Street %d", i),
				fmt.Sprintf("City %d", i),
				fmt.Sprintf("%d", i),
				fmt.Sprintf("Cust.omer%d@email.com", i),
				fmt.Sprintf("%d", i),
			},
		}
	}
	return items
}

func main() {
	tmpl, err := template.ParseFiles("../../../List.html")
	if err != nil {
		fmt.Println(err)
		return
	}

	items := GenerateRange(1000)

	var times = make([]int64, RUNS)
	for i := 0; i < RUNS; i++ {
		var w bytes.Buffer

		start := time.Now()
		tmpl.ExecuteTemplate(&w, "table", items)
		duration := time.Since(start)
		times[i] = duration.Microseconds()
	}

	var sum int64
	var min int64
	var max int64

	sum = 0
	min = times[0]
	max = 0

	for i := 0; i < RUNS; i++ {
		if times[i] < min {
			min = times[i]
		}
		if times[i] > max {
			max = times[i]
		}
		sum += times[i]
	}

	average := sum / RUNS

	fmt.Printf("Table with 1000 items, %d runs:\n", RUNS)
	fmt.Printf("Min: %d (micros)\n", min)
	fmt.Printf("Max: %d (micros)\n", max)
	fmt.Printf("Average: %d (micros)\n", average)

}
